akka-actor/src/main/scala/akka/util/ImmutableIntMap.scala

```diff
@@ -3,139 +3,123 @@
  */
 package akka.util
 import java.util.Arrays
-
 import akka.annotation.InternalApi
-
 import scala.annotation.tailrec
 
 /**
  * INTERNAL API
  */
 @InternalApi private[akka] object ImmutableIntMap {
-  val empty: ImmutableIntMap =
-    new ImmutableIntMap(Array.emptyIntArray, Array.empty)
-
-  private final val MaxScanLength = 10
+  final val empty: ImmutableIntMap = new ImmutableIntMap(Array.emptyLongArray)
 }
 
 /**
  * INTERNAL API
- * Specialized Map for primitive `Int` keys to avoid allocations (boxing).
- * Keys and values are backed by arrays and lookup is performed with binary
- * search. It's intended for rather small (<1000) maps.
+ * Specialized Map for primitive `Int` keys and values to avoid allocations (boxing).
+ * Keys and values are encoded in a single Long array and does copy-on-write with no
+ * structural sharing, it's intended for rather small maps (<1000 elements).
  */
-@InternalApi private[akka] final class ImmutableIntMap private (
-  private val keys: Array[Int], private val values: Array[Int]) {
+@InternalApi private[akka] final class ImmutableIntMap private (private final val kvs: Array[Long]) {
+  final val size: Int = kvs.length
+
+  private[this] final def indexForKey(key: Int): Int = {
+    @tailrec def find(lo: Int, hi: Int): Int =
+      if (lo <= hi) {
+        val at = (lo + hi) >>> 1
+        val eKey = keyFromEntry(kvs(at))
+        if (eKey < key) find(at + 1, hi)
+        else if (eKey > key) find(lo, at - 1)
+        else at
+      } else -(lo + 1)
+    find(0, size - 1)
+  }
 
-  final val size: Int = keys.length
+  @inline private[this] final def entry(key: Int, value: Int): Long = (key.toLong << 32) | (value & 0xffffffffL)
+  @inline private[this] final def keyFromEntry(entry: Long): Int = (entry >> 32).toInt // get rid of value bits
+  @inline private[this] final def valueFromEntry(entry: Long): Int = entry.toInt //cast away key bits
 
   /**
    * Worst case `O(log n)`, allocation free.
+   * Will return Int.MinValue if not found, so beware of storing Int.MinValues
    */
-  def get(key: Int): Int = {
-    val i = Arrays.binarySearch(keys, key)
-    if (i >= 0) values(i)
-    else Int.MinValue // cant use null, cant use OptionVal, other option is to throw an exception...
+  final def get(key: Int): Int = {
+    @tailrec def find(lo: Int, hi: Int): Int =
+      if (lo <= hi) {
+        val at = (lo + hi) >>> 1
```

Comment:
Yep. https://research.googleblog.com/2006/06/extra-extra-read-all-about-it-nearly.html